File: crypto_code/LAB3/AESClient.java
import java.net.*;
import java.io.*;
import java.util.Scanner;

public class AESClient {
    public static void main(String[] args) throws Exception {

        Scanner sc = new Scanner(System.in);
        System.out.print("Enter Plaintext (max 16 characters): ");
        String pt = sc.nextLine();

        // Pad or truncate to 16 characters
        while (pt.length() < 16) pt += "X";
        pt = pt.substring(0, 16);

        System.out.print("Enter 16-character Key: ");
        String keyStr = sc.nextLine();
        while (keyStr.length() < 16) keyStr += "0";
        keyStr = keyStr.substring(0, 16);

        System.out.println("\n========== AES-128 ENCRYPTION ==========\n");
        System.out.println("Plaintext: " + pt);
        System.out.println("Key: " + keyStr);
        System.out.println();

        // Perform full encryption with round-by-round output
        String ciphertext = AESCore.encrypt(pt, keyStr);

        System.out.println("========== AES-128 DECRYPTION ==========\n");
        System.out.println("Ciphertext to Decrypt: " + ciphertext);
        System.out.println();

        // Perform full decryption with round-by-round output
        String decrypted = AESCore.decrypt(ciphertext, keyStr);

        System.out.println("========== VERIFICATION ==========");
        System.out.println("Original Plaintext: " + pt);
        System.out.println("Decrypted Result:   " + decrypted);
        
        // Convert hex to ASCII for better readability
        String decryptedASCII = hexToASCII(decrypted);
        System.out.println("Decrypted (ASCII): " + decryptedASCII);
        System.out.println("Match: " + pt.equals(decryptedASCII));

        // Send ciphertext to server
        try {
            Socket s = new Socket("localhost", 8060);
            PrintWriter out = new PrintWriter(s.getOutputStream(), true);
            System.out.println("\n========== SENDING TO SERVER ==========");
            out.println(ciphertext);
            System.out.println("Ciphertext sent to server: " + ciphertext);
            s.close();
        } catch (Exception e) {
            System.out.println("Server not available - skipping network transmission");
        }

        sc.close();
    }

    private static String hexToASCII(String hex) {
        StringBuilder result = new StringBuilder();
        for (int i = 0; i < hex.length(); i += 2) {
            String hexByte = hex.substring(i, i + 2);
            int charCode = Integer.parseInt(hexByte, 16);
            result.append((char) charCode);
        }
        return result.toString();
    }
}

File: crypto_code/LAB3/AESClient.py
#!/usr/bin/env python3
"""
AES Client - Encrypts plaintext and sends it to server
Server: localhost:8060
"""

import socket
from AES import encrypt

HOST = '127.0.0.1'
PORT = 8060

print("========== AES CLIENT ==========\n")

# Get plaintext from user
plaintext = input("Enter Plaintext (max 16 characters): ").strip()

if not plaintext:
    plaintext = "HELLOWORLD12345"
    print(f"Using default plaintext: {plaintext}")

# Pad or truncate to 16 characters
plaintext = (plaintext + "X" * 16)[:16]

# Get key from user
key_str = input("Enter 16-character Key: ").strip()

if not key_str:
    key_str = "mysecretkey1234"
    print(f"Using default key: {key_str}")

# Pad or truncate key to 16 characters
key_str = (key_str + "\0" * 16)[:16]

print("\n========== AES-128 ENCRYPTION ==========")
print(f"\nPlaintext: {plaintext}")
print(f"Key: {key_str}\n")

# Encrypt the plaintext
ciphertext = encrypt(plaintext, key_str)

print(f"\n========== ENCRYPTION RESULT ==========")
print(f"Final Ciphertext (Hex): {ciphertext}\n")

# Send to server
print("========== SENDING TO SERVER ==========")

try:
    # Create TCP socket
    client_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    client_socket.connect((HOST, PORT))
    
    # Send ciphertext
    client_socket.sendall(ciphertext.encode())
    print(f"Ciphertext sent to server: {ciphertext}")
    
    # Receive decrypted plaintext from server
    response = client_socket.recv(1024).decode()
    print(f"\nServer says: {response}\n")
    
    client_socket.close()
    
except ConnectionRefusedError:
    print(f"Error: Could not connect to server on {HOST}:{PORT}")
    print("Please make sure AESServer.py is running.")
except Exception as e:
    print(f"Error: {e}")
